# REDES NEURAIS ARTIFICIAIS

# PARTE 1: Classificação Binária - Diagnóstico de Câncer de Mama

Base de dados: Breast Cancer Wisconsin (Diagnostic) -
https://archive.ics.uci.edu/dataset/17/breast+cancer+wisconsin+diagnostic

In [291]:
#bilbiotecas keras e tensorflow
#!pip install scikeras==0.13.0

In [292]:
#!pip install tensorflow==2.16.1 scikit-learn==1.5.0

In [293]:
# após instalar as bibliotecas, é necessário reiniciar o notebook
# Para isso, rode esta célula sozinha antes de prosseguir.
#exit()

In [294]:
import pandas as pd
import tensorflow as tf
import sklearn
import scikeras

In [295]:
pd.__version__, tf.__version__, sklearn.__version__, scikeras.__version__

('2.3.3', '2.16.1', '1.5.0', '0.13.0')

### Carregar os dados

1) Carregue a base de dados, faça a divisão de treino e teste (para isso, utilize a função train_test_split do sklearn), como o tamanho da base de teste de 0.25.

In [296]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

# Carregar os dados
X = pd.read_csv('dados_breast.csv').values
y = pd.read_csv('rotulos_breast.csv', header=None).values.ravel()

# Garantir que X e y tenham o mesmo tamanho (ajustar y para o tamanho de X)
min_size = min(len(X), len(y))
X = X[:min_size]
y = y[:min_size]

# inicialmente os arquivos tinham uma diferença de tamanho (um era maior que o outro)
print(f"Tamanho de X: {X.shape}")
print(f"Tamanho de y: {y.shape}")

# Dividir em treino e teste (75% treino, 25% teste)
X_treinamento, X_teste, y_treinamento, y_teste = train_test_split(X, y, test_size=0.25, random_state=42)

Tamanho de X: (569, 30)
Tamanho de y: (569,)


In [297]:
X_treinamento.shape, y_treinamento.shape

((426, 30), (426,))

In [298]:
X_teste.shape, y_teste.shape

((143, 30), (143,))

### Estrutura da Rede Neural Artificial e Teste

In [299]:
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import cross_val_score
from tensorflow.keras.models import Sequential
import tensorflow as tf
from tensorflow.keras import backend as k

2) Crie a RNA com as seguintes configurações:

a) uma camada de entrada de com 30 neurônios;

b) uma camada oculta densa com 16 neurônios. Porque a camada oculta tem essa quantidade de neurônios?

c) adicione na camada oculta a função de ativação relu e inicialize os pesos de utilize o Random uniform initializer (https://www.tensorflow.org/api_docs/python/tf/keras/initializers);

d) adicione a camada de saíde com a função de ativação sigmoid. Porque foi utilizada esta função para a saída da RNA?

3) Porque utilizamos a classe chamada Sequential para a RNA?

In [300]:
# Criar a Rede Neural Artificial
rede_neural = Sequential([
    # a) Camada de entrada com 30 neurônios (30 features do dataset)
    tf.keras.layers.InputLayer(shape=(30,)),

    # b) e c) Camada oculta densa com 16 neurônios, ativação relu e inicializador random_uniform
    tf.keras.layers.Dense(units=16, activation='relu', kernel_initializer='random_uniform'),

    # d) Camada de saída com 1 neurônio e ativação sigmoid (classificação binária)
    tf.keras.layers.Dense(units=1, activation='sigmoid')
])

**Respostas:**

**b)** A camada oculta tem 16 neurônios como uma escolha heurística que equilibra capacidade de aprendizado e complexidade. Geralmente, usa-se um número entre a quantidade de entradas (30) e saídas (1). O valor 16 é suficiente para capturar padrões nos dados sem causar overfitting excessivo.

**d)** A função sigmoid foi utilizada na saída porque estamos fazendo classificação binária (0 ou 1: benigno ou maligno). A sigmoid transforma qualquer valor em um número entre 0 e 1, que pode ser interpretado como probabilidade da amostra pertencer à classe positiva.

**3)** Utilizamos a classe Sequential porque ela permite criar modelos de forma linear e sequencial, onde as camadas são empilhadas uma após a outra. É ideal para redes neurais feedforward simples onde cada camada tem exatamente uma entrada e uma saída.

4) A partir da RNA gerada, explique o que são os valores apresentados na tabela da rede_neural.summary()

In [301]:
rede_neural.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 513 (2.00 KB)

 Trainable params: 513 (2.00 KB)

 Non-trainable params: 0 (0.00 B)

**Resposta 4:** A tabela do summary() mostra:
- **Layer (type)**: O tipo de cada camada (InputLayer, Dense)
- **Output Shape**: Formato da saída de cada camada (None, número_neurônios), onde None representa o batch size
- **Param #**: Número de parâmetros treináveis em cada camada. Para camadas Dense, é calculado como: (neurônios_entrada × neurônios_saída) + neurônios_saída (bias)
  - Primeira camada oculta: (30 × 16) + 16 = 496 parâmetros
  - Camada de saída: (16 × 1) + 1 = 17 parâmetros
  - **Total**: 513 parâmetros treináveis

5) Adicione um otimizador Adam (https://www.tensorflow.org/api_docs/python/tf/keras/optimizers) e especifique a classe loss binário - binary crossentropy (https://www.tensorflow.org/api_docs/python/tf/keras/losses) e a classe metrics para utilizar a métrica de avaliação de acurácia binária (https://www.tensorflow.org/api_docs/python/tf/keras/metrics)

6) Para que servem os otimizadores? Como o otimizador Adam funciona?

7) Depois de estruturados os parâmtros da RNA, utilize a função .fit para fazer o treinamento da rede. Como foi utilizado o otimizador Adam, e ele é baseado na descida do gradiente estocástica, é possível definir a quantidade de registros que serão enviados para a RNA, isto é, em cada batch serão utilizados 10 registros. Quantos batches serão utilizados ao total?

8) Por fim, defina o número de épocas em que ocorre o treinamento igual a 100.

In [302]:
# 5) Compilar a rede neural com otimizador Adam, loss binary_crossentropy e métrica binary_accuracy
rede_neural.compile(optimizer='adam', loss='binary_crossentropy', metrics=['binary_accuracy'])

**Respostas:**

**6)** Os otimizadores servem para ajustar os pesos da rede neural durante o treinamento, minimizando a função de perda. O Adam (Adaptive Moment Estimation) combina as vantagens de dois outros otimizadores: RMSprop e Momentum. Ele adapta a taxa de aprendizado individualmente para cada parâmetro, usando médias móveis dos gradientes e dos quadrados dos gradientes, tornando-o eficiente e convergindo rapidamente.

**7)** Com batch_size=10 e 426 amostras de treino (75% de 569), teremos aproximadamente 43 batches por época (426 ÷ 10 = 42.6, arredondado para 43).

In [303]:
# 7) e 8) Treinar a rede neural com batch_size=10 e epochs=100
rede_neural.fit(X_treinamento, y_treinamento, batch_size=10, epochs=100)

Epoch 1/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - binary_accuracy: 0.5587 - loss: 12.1746
Epoch 2/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5469 - loss: 3.9643
Epoch 3/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5540 - loss: 3.5066
Epoch 4/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5610 - loss: 2.8726
Epoch 5/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - binary_accuracy: 0.5329 - loss: 4.0444
Epoch 6/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5681 - loss: 3.3538
Epoch 7/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - binary_accuracy: 0.5399 - loss: 2.9036
Epoch 8/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.5469 - loss: 2.7278
Epoch 9/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5798 - loss: 2.6910
Epoch 10/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5728 - loss: 2.5262
Epoch 11/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.6056 

9) Crie uma variável chamada previsoes para realizar a previsão dos dados de teste (X_teste) O resultado da rede deve ser um valor entre 0 e 1. Porque isso acontece?

In [304]:
previsoes = rede_neural.predict(X_teste)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


In [305]:
# Visualizar as previsões
print("Primeiras 10 previsões:")
print(previsoes[:10])
print(f"\nValor mínimo: {previsoes.min():.4f}")
print(f"Valor máximo: {previsoes.max():.4f}")
print(f"Valor médio: {previsoes.mean():.4f}")
print(f"Todas as previsões estão entre 0 e 1: {(previsoes >= 0).all() and (previsoes <= 1).all()}")

Primeiras 10 previsões:
[[0.9714977 ]
 [0.9755433 ]
 [0.90687454]
 [0.7719738 ]
 [0.97718644]
 [0.99078125]
 [0.9969329 ]
 [0.9680954 ]
 [0.95538265]
 [0.8787439 ]]

Valor mínimo: 0.0000
Valor máximo: 1.0000
Valor médio: 0.9126
Todas as previsões estão entre 0 e 1: True


**Resposta 9:** O resultado está entre 0 e 1 porque a camada de saída usa a função de ativação sigmoid, que matematicamente transforma qualquer valor de entrada em um valor entre 0 e 1. Isso representa a probabilidade da amostra pertencer à classe positiva (maligno = 1). Valores próximos a 1 indicam alta probabilidade de ser maligno, e valores próximos a 0 indicam alta probabilidade de ser benigno.

10) Converta o resultado que está em probabilidade para valores binários com um limiar (treshold) de 0.5.

In [306]:
previsoes = previsoes > 0.5

In [307]:
previsoes

array([[ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [ True],
       [

9) Explique o resultado da RNA.

In [308]:
# Verificar a distribuição REAL dos dados
print("Distribuição REAL no dataset completo:")
print(f"Benignos (0): {(y == 0).sum()} ({(y == 0).mean():.2%})")
print(f"Malignos (1): {(y == 1).sum()} ({(y == 1).mean():.2%})")

print("\nDistribuição no conjunto de TESTE:")
print(f"Benignos (0): {(y_teste == 0).sum()} ({(y_teste == 0).mean():.2%})")
print(f"Malignos (1): {(y_teste == 1).sum()} ({(y_teste == 1).mean():.2%})")

# Visualizar as previsões binárias
print("Primeiras 10 previsões:")
print(previsoes[:10].flatten())  # Mostra True/False

# Estatísticas para dados binários
previsoes_num = previsoes.astype(int).flatten()
print(f"\nTotal de predições classe 0 (benigno): {(previsoes_num == 0).sum()}")
print(f"Total de predições classe 1 (maligno): {(previsoes_num == 1).sum()}")
print(f"\nPorcentagem real de malignos no teste: {(y_teste == 1).mean():.2%}")
print(f"Proporção de casos malignos: {previsoes_num.mean():.2%}")


Distribuição REAL no dataset completo:
Benignos (0): 213 (37.43%)
Malignos (1): 356 (62.57%)

Distribuição no conjunto de TESTE:
Benignos (0): 61 (42.66%)
Malignos (1): 82 (57.34%)
Primeiras 10 previsões:
[ True  True  True  True  True  True  True  True  True  True]

Total de predições classe 0 (benigno): 5
Total de predições classe 1 (maligno): 138

Porcentagem real de malignos no teste: 57.34%
Proporção de casos malignos: 96.50%


**Resposta:** O método `evaluate()` retorna dois valores: a perda (loss) e a acurácia binária. A perda indica o quão distantes estão as previsões dos valores reais (quanto menor, melhor). A acurácia binária mostra a porcentagem de classificações corretas no conjunto de teste. Esses resultados permitem avaliar o desempenho do modelo em dados não vistos durante o treinamento.

## Camadas e Otimização da RNA

10) Adicione mais uma camada para oculta densa com 16 neurônio com a função de ativação relu e inicialize os pesos de utilize o Random uniform initializer. Qual o total de parâmetros da RNA agora?

In [309]:
# Recriar a rede neural com duas camadas ocultas
rede_neural = Sequential([
    tf.keras.layers.InputLayer(shape=(30,)),
    tf.keras.layers.Dense(units=16, activation='relu', kernel_initializer='random_uniform'),
    tf.keras.layers.Dense(units=16, activation='relu', kernel_initializer='random_uniform'),  # Nova camada
    tf.keras.layers.Dense(units=1, activation='sigmoid')
])

In [310]:
rede_neural.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_5 (Dense)                 │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 16)             │           272 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 785 (3.07 KB)

 Trainable params: 785 (3.07 KB)

 Non-trainable params: 0 (0.00 B)

11) Abaixo são adicionados os parâmetros do otimizador, que são a taxa de aprendizado e o clipvalue. O que eles fazem?

In [311]:
otimizador = tf.keras.optimizers.Adam(learning_rate = 0.001, clipvalue = 0.5)
rede_neural.compile(optimizer = otimizador, loss = 'binary_crossentropy', metrics = ['binary_accuracy'])

**Resposta 11:**
- **learning_rate (taxa de aprendizado)**: Controla o tamanho dos passos que o otimizador dá ao ajustar os pesos. Um valor de 0.001 significa passos pequenos e cuidadosos, evitando oscilações mas permitindo convergência gradual.
- **clipvalue**: Limita os gradientes ao intervalo [-0.5, 0.5]. Isso previne o problema de "gradientes explosivos" (exploding gradients), onde gradientes muito grandes podem desestabilizar o treinamento.

12) Teste novamente a RNA. Aumentar a quantidade de camadas melhorou ou piorou os resultados? Explique o que aconteceu com a RNA e porque.

In [312]:
rede_neural.fit(X_treinamento, y_treinamento, batch_size = 10, epochs = 100)

Epoch 1/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - binary_accuracy: 0.5235 - loss: 1.3475
Epoch 2/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5798 - loss: 1.0237
Epoch 3/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5775 - loss: 0.9652
Epoch 4/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5352 - loss: 1.0682
Epoch 5/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6080 - loss: 0.9422
Epoch 6/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6009 - loss: 0.8685
Epoch 7/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5775 - loss: 0.8505
Epoch 8/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5845 - loss: 1.1048
Epoch 9/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5986 - loss: 1.0210
Epoch 10/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6221 - loss: 0.9514 
Epoch 11/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6408 

In [313]:
previsoes = rede_neural.predict(X_teste)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step


In [314]:
rede_neural.evaluate(X_teste, y_teste)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - binary_accuracy: 0.5944 - loss: 0.8634  


[0.8633701801300049, 0.5944055914878845]

In [315]:
# Verificar a distribuição REAL dos dados
print("Distribuição REAL no dataset completo:")
print(f"Benignos (0): {(y == 0).sum()} ({(y == 0).mean():.2%})")
print(f"Malignos (1): {(y == 1).sum()} ({(y == 1).mean():.2%})")

print("\nDistribuição no conjunto de TESTE:")
print(f"Benignos (0): {(y_teste == 0).sum()} ({(y_teste == 0).mean():.2%})")
print(f"Malignos (1): {(y_teste == 1).sum()} ({(y_teste == 1).mean():.2%})")

# Visualizar as probabilidades ANTES da conversão
print("\nPrimeiras 10 probabilidades:")
print(previsoes[:10].flatten())

# CONVERTER para binário (0 ou 1) usando threshold de 0.5
previsoes_binarias = (previsoes > 0.5).astype(int).flatten()

# Visualizar as previsões APÓS conversão
print("\nPrimeiras 10 previsões binárias:")
print(previsoes_binarias[:10])

# Estatísticas das PREDIÇÕES
print(f"\nTotal de predições classe 0 (benigno): {(previsoes_binarias == 0).sum()}")
print(f"Total de predições classe 1 (maligno): {(previsoes_binarias == 1).sum()}")
print(f"Proporção de casos malignos preditos: {previsoes_binarias.mean():.2%}")

# Comparar com a realidade
print(f"\nPorcentagem real de malignos no teste: {(y_teste == 1).mean():.2%}")
print(f"Porcentagem predita de malignos: {previsoes_binarias.mean():.2%}")


Distribuição REAL no dataset completo:
Benignos (0): 213 (37.43%)
Malignos (1): 356 (62.57%)

Distribuição no conjunto de TESTE:
Benignos (0): 61 (42.66%)
Malignos (1): 82 (57.34%)

Primeiras 10 probabilidades:
[0.7341141  0.637027   0.54497194 0.42876038 0.6386595  0.8447914
 0.9908162  0.65471387 0.61306256 0.5894706 ]

Primeiras 10 previsões binárias:
[1 1 1 0 1 1 1 1 1 1]

Total de predições classe 0 (benigno): 21
Total de predições classe 1 (maligno): 122
Proporção de casos malignos preditos: 85.31%

Porcentagem real de malignos no teste: 57.34%
Porcentagem predita de malignos: 85.31%


## K-Fold Cross Validation

13) Assista o vídeo https://youtu.be/RczbeFs_WbQ?si=f1Yu4LZaGDfZZpbZ para compreender melhor sobre como funciona o k-fold cross validation. Explique como esta rede foi configurada e como é possível chegar no resultado dela. O que é necessário fazer?

**Resposta 13:** O K-Fold Cross Validation divide os dados em K partes (folds). Para cada iteração, uma parte é usada como teste e as demais como treino, repetindo K vezes. A rede é configurada através da função `criar_rede()` que define a arquitetura, e o `cross_val_score` treina e avalia K vezes (cv=10), retornando a acurácia de cada fold. Isso permite uma avaliação mais robusta do modelo.

14) Calcule também o Desvio Padrão dos resultados para avaliar o modelo. O que é possível concluir com esse resultado?

**Resposta 14:** O desvio padrão de 3.88% indica consistência nos resultados entre os folds, porém a acurácia de 61.17% é muito baixa, sugerindo underfitting (modelo não aprende adequadamente os padrões dos dados). São necessárias melhorias como normalização dos dados e ajuste de hiperparâmetros.



In [316]:
# Calcular média e desvio padrão dos resultados
media = resultados.mean()
desvio_padrao = resultados.std()
print(f"Acurácia Média: {media:.4f}")
print(f"Desvio Padrão: {desvio_padrao:.4f}")

Acurácia Média: 0.6257
Desvio Padrão: 0.0314


In [317]:
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import cross_val_score
from tensorflow.keras.models import Sequential
from tensorflow.keras import backend as k

In [318]:
def criar_rede():
  k.clear_session()
  rede_neural = Sequential([
      tf.keras.layers.InputLayer(shape=(30,)),
      tf.keras.layers.Dense(units=16, activation='relu', kernel_initializer='random_uniform'),
      tf.keras.layers.Dense(units=16, activation='relu', kernel_initializer='random_uniform'),
      tf.keras.layers.Dense(units=1, activation = 'sigmoid')])
  otimizador = tf.keras.optimizers.Adam(learning_rate = 0.001, clipvalue = 0.5)
  rede_neural.compile(optimizer = otimizador, loss = 'binary_crossentropy', metrics = ['binary_accuracy'])
  return rede_neural

In [319]:
rede_neural = KerasClassifier(model = criar_rede, epochs = 100, batch_size = 10)

In [320]:
resultados = cross_val_score(estimator = rede_neural, X = X, y = y, cv = 10, scoring = 'accuracy')

Epoch 1/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - binary_accuracy: 0.5586 - loss: 1.2342
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5820 - loss: 0.9298
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5859 - loss: 0.8815
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5801 - loss: 0.9514
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5996 - loss: 0.7934
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5957 - loss: 0.7912
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6230 - loss: 0.8603
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6211 - loss: 0.7937
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5977 - loss: 0.7644
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5957 - loss: 0.7875
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6172 -

In [321]:
resultados

array([0.42105263, 0.54385965, 0.54385965, 0.56140351, 0.50877193,
       0.56140351, 0.54385965, 0.61403509, 0.54385965, 0.66071429])

In [322]:
# Calcular média e desvio padrão dos resultados
media = resultados.mean()
desvio_padrao = resultados.std()
print(f"Acurácia Média: {media:.4f}")
print(f"Desvio Padrão: {desvio_padrao:.4f}")

Acurácia Média: 0.5503
Desvio Padrão: 0.0592


## Overfitting e Dropout

14) Como vimos, o modelo anterior sofreu overfitting. Para reduzir esse problema, aplicamos a técnica de regularização conhecida como dropout (artigo oficial: https://www.cs.toronto.edu/~rsalakhu/papers/srivastava14a.pdf e explicação em português: https://www.deeplearningbook.com.br/como-funciona-o-dropout/ ). Aplique o dropout de 20% na primeira e segunda camada oculta. O que acontece com os resultados? E o Desvio Padrão?

In [323]:
def criar_rede():
  k.clear_session()
  rede_neural = Sequential([
      tf.keras.layers.InputLayer(shape=(30,)),
      tf.keras.layers.Dense(units=16, activation='relu', kernel_initializer='random_uniform'),
      tf.keras.layers.Dropout(rate=0.2),  # Dropout de 20% após a primeira camada oculta
      tf.keras.layers.Dense(units=16, activation='relu', kernel_initializer='random_uniform'),
      tf.keras.layers.Dropout(rate=0.2),  # Dropout de 20% após a segunda camada oculta
      tf.keras.layers.Dense(units=1, activation = 'sigmoid')])
  otimizador = tf.keras.optimizers.Adam(learning_rate = 0.001, clipvalue = 0.5)
  rede_neural.compile(optimizer = otimizador, loss = 'binary_crossentropy', metrics = ['binary_accuracy'])
  return rede_neural

In [324]:
rede_neural = KerasClassifier(model = criar_rede, epochs = 100, batch_size = 10)

In [325]:
resultados = cross_val_score(estimator = rede_neural, X = X, y = y, cv = 10, scoring = 'accuracy')

Epoch 1/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - binary_accuracy: 0.5527 - loss: 1.6470
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5879 - loss: 0.8950
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5527 - loss: 0.8788
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5879 - loss: 0.7955
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6172 - loss: 0.7619
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6016 - loss: 0.7863
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - binary_accuracy: 0.6074 - loss: 0.7299
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - binary_accuracy: 0.6250 - loss: 0.7881
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - binary_accuracy: 0.6172 - loss: 0.7541
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6113 - loss: 0.7567
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6465 -

In [326]:
resultados

array([0.63157895, 0.64912281, 0.66666667, 0.61403509, 0.63157895,
       0.57894737, 0.64912281, 0.57894737, 0.64912281, 0.60714286])

In [327]:
# Calcular média e desvio padrão dos resultados com dropout
media_dropout = resultados.mean()
desvio_padrao_dropout = resultados.std()
print(f"Acurácia Média (com Dropout): {media_dropout:.4f}")
print(f"Desvio Padrão (com Dropout): {desvio_padrao_dropout:.4f}")

Acurácia Média (com Dropout): 0.6256
Desvio Padrão (com Dropout): 0.0287


**Resposta 14:** O dropout ajuda a reduzir o overfitting ao "desligar" aleatoriamente 20% dos neurônios durante o treinamento, forçando a rede a aprender representações mais robustas. Geralmente, isso resulta em uma acurácia de treino ligeiramente menor, mas uma acurácia de teste melhor (melhor generalização). O desvio padrão tende a diminuir, indicando que o modelo é mais consistente entre diferentes folds.

## Tuning dos Hiperparâmetros (técnicas de ajuste)

Hyperparameter Tuning (Ajuste de Hiperparâmetros) é o processo geral de encontrar os melhores valores para os hiperparâmetros de um modelo de machine learning.
Exemplos de hiperparâmetros:

- Taxa de aprendizado (learning rate)

- Número de árvores em um RandomForest

- Número de camadas ou neurônios em uma rede neural

- Tipo de kernel em um SVM

O tuning pode ser feito por diferentes técnicas, como:

- Grid Search

- Random Search

- Bayesian Optimization, etc.

15) Descreva como a RNA foi configurada para fazer o processo de tuning.

**Resposta 15:** A RNA foi configurada de forma parametrizada através da função `criar_rede()`, que agora aceita múltiplos parâmetros (optimizer, loss, kernel_initializer, activation, neurons). Esses parâmetros são definidos no dicionário `parametros` com diferentes valores a serem testados. O GridSearchCV testa todas as combinações possíveis desses parâmetros, treinando e avaliando o modelo para cada combinação, retornando os melhores parâmetros encontrados.

In [328]:
from sklearn.model_selection import GridSearchCV

def criar_rede(optimizer, loss, kernel_initializer, activation, neurons):
  k.clear_session()
  rede_neural = Sequential([
      tf.keras.layers.InputLayer(shape=(30,)),
      tf.keras.layers.Dense(units=neurons, activation=activation, kernel_initializer=kernel_initializer),
      tf.keras.layers.Dropout(rate = 0.2),
      tf.keras.layers.Dense(units=neurons, activation=activation, kernel_initializer=kernel_initializer),
      tf.keras.layers.Dropout(rate = 0.2),
      tf.keras.layers.Dense(units=1, activation = 'sigmoid')])
  rede_neural.compile(optimizer = optimizer, loss = loss, metrics = ['binary_accuracy'])
  return rede_neural

In [329]:
rede_neural = KerasClassifier(model = criar_rede)

In [330]:
parametros = {
    'batch_size': [10, 30],
    'epochs': [50, 100],
    'model__optimizer': ['adam', 'sgd'],
    'model__loss': ['binary_crossentropy', 'hinge'],
    'model__kernel_initializer': ['random_uniform', 'normal'],
    'model__activation': ['relu', 'tanh'],
    'model__neurons': [16, 8]
}

In [331]:
#caso tiver problema com a configuração anterior, testar essa
parametros = {
    'batch_size': [10, 30],
    'epochs': [50],
    'model__optimizer': ['adam'],
    'model__loss': ['binary_crossentropy'],
    'model__kernel_initializer': ['random_uniform', 'normal'],
    'model__activation': ['relu'],
    'model__neurons': [16]
}

In [332]:
parametros

{'batch_size': [10, 30],
 'epochs': [50],
 'model__optimizer': ['adam'],
 'model__loss': ['binary_crossentropy'],
 'model__kernel_initializer': ['random_uniform', 'normal'],
 'model__activation': ['relu'],
 'model__neurons': [16]}

In [333]:
grid_search = GridSearchCV(estimator = rede_neural, param_grid = parametros,
                          scoring = 'accuracy', cv = 5)

In [334]:
grid_search = grid_search.fit(X, y)

Epoch 1/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - binary_accuracy: 0.5297 - loss: 1.7022
Epoch 2/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5824 - loss: 0.9001
Epoch 3/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - binary_accuracy: 0.6176 - loss: 0.8151
Epoch 4/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6110 - loss: 0.7746
Epoch 5/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - binary_accuracy: 0.6330 - loss: 0.7346
Epoch 6/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5846 - loss: 0.7597
Epoch 7/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - binary_accuracy: 0.6286 - loss: 0.6918
Epoch 8/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - binary_accuracy: 0.6374 - loss: 0.6669
Epoch 9/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6242 - loss: 0.6806
Epoch 10/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - binary_accuracy: 0.6242 - loss: 0.6709 
Epoch 11/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6286 - loss: 0.6

In [335]:
print(grid_search)

GridSearchCV(cv=5,
             estimator=KerasClassifier(model=<function criar_rede at 0x0000029F95D53740>),
             param_grid={'batch_size': [10, 30], 'epochs': [50],
                         'model__activation': ['relu'],
                         'model__kernel_initializer': ['random_uniform',
                                                       'normal'],
                         'model__loss': ['binary_crossentropy'],
                         'model__neurons': [16], 'model__optimizer': ['adam']},
             scoring='accuracy')


In [336]:
melhores_parametros = grid_search.best_params_
print(melhores_parametros)

{'batch_size': 30, 'epochs': 50, 'model__activation': 'relu', 'model__kernel_initializer': 'random_uniform', 'model__loss': 'binary_crossentropy', 'model__neurons': 16, 'model__optimizer': 'adam'}


In [337]:
melhor_precisao = grid_search.best_score_
print(melhor_precisao)

0.6204005589194225


16) É possível melhorar ainda mais a curácia da RNA? Como?

**Resposta 16:** Sim, é possível melhorar a acurácia através de várias técnicas:
- **Normalização/Padronização dos dados**: Usar StandardScaler ou MinMaxScaler para escalar as features
- **Aumentar épocas**: Treinar por mais tempo, com early stopping para evitar overfitting
- **Ajustar arquitetura**: Testar diferentes números de camadas e neurônios
- **Regularização L1/L2**: Adicionar penalização aos pesos
- **Batch Normalization**: Normalizar as saídas entre camadas
- **Ajustar learning rate**: Usar learning rate schedulers ou valores diferentes
- **Aumentar dados**: Data augmentation (se aplicável) ou coletar mais amostras
- **Ensemble methods**: Combinar múltiplos modelos

#Para salvar o seu modelo de RNA

## Melhorias da RNA (Normalização, Regularização, BatchNorm, EarlyStopping)

**Seção de Melhorias**
Aplicaremos algumas das estratégias mencionadas na Resposta 16 para tentar melhorar a curácia e a generalização:

1. Normalização (StandardScaler) dos atributos antes de treinar (via Pipeline).
2. Aumento do número de épocas com EarlyStopping (evita overfitting e treinos desnecessários).
3. Regularização L2 nos pesos das camadas densas (reduz overfitting).
4. Batch Normalization para estabilizar e acelerar o treinamento.
5. Ajuste de learning rate (menor) para passos mais precisos.

Depois, recalculamos `resultados` (K-Fold) e comparamos média e desvio padrão.

In [338]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

# Callback de EarlyStopping
early_stopping = tf.keras.callbacks.EarlyStopping(monitor='loss', patience=10, restore_best_weights=True)

def criar_rede_melhorada():
    k.clear_session()
    reg = tf.keras.regularizers.l2(0.001)
    model = Sequential([
        tf.keras.layers.InputLayer(shape=(30,)),
        # Camada 1
        tf.keras.layers.Dense(16, kernel_initializer='random_uniform', kernel_regularizer=reg, use_bias=True),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Activation('relu'),
        tf.keras.layers.Dropout(0.2),
        # Camada 2
        tf.keras.layers.Dense(16, kernel_initializer='random_uniform', kernel_regularizer=reg, use_bias=True),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Activation('relu'),
        tf.keras.layers.Dropout(0.2),
        # Camada 3 (menor para compactar representação)
        tf.keras.layers.Dense(8, kernel_initializer='random_uniform', kernel_regularizer=reg, use_bias=True),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Activation('relu'),
        # Saída
        tf.keras.layers.Dense(1, activation='sigmoid')
    ])
    # Learning rate menor para treinamento mais suave
    opt = tf.keras.optimizers.Adam(learning_rate=0.0005, clipvalue=0.5)
    model.compile(optimizer=opt, loss='binary_crossentropy', metrics=['binary_accuracy'])
    return model

# KerasClassifier atualizado
rede_neural_melhorada = KerasClassifier(
    model=criar_rede_melhorada,
    epochs=150,            # mais épocas (EarlyStopping interrompe antes se necessário)
    batch_size=10,
    callbacks=[early_stopping]
)

# Pipeline com normalização + modelo
pipeline_melhorado = Pipeline([
    ('scaler', StandardScaler()),
    ('nn', rede_neural_melhorada)
])

# Recalcular resultados com melhorias (cv=10 para manter consistência)
resultados = cross_val_score(estimator=pipeline_melhorado, X=X, y=y, cv=10, scoring='accuracy')
print("Acurácias dos 10 folds (melhorado):", resultados)

media = resultados.mean()
desvio_padrao = resultados.std()
print(f"Acurácia Média (melhorado): {media:.4f}")
print(f"Desvio Padrão (melhorado): {desvio_padrao:.4f}")

Epoch 1/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - binary_accuracy: 0.4629 - loss: 0.7426
Epoch 2/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5488 - loss: 0.6899
Epoch 3/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.5352 - loss: 0.7080
Epoch 4/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6094 - loss: 0.6710
Epoch 5/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6406 - loss: 0.6639
Epoch 6/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5996 - loss: 0.6700
Epoch 7/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6230 - loss: 0.6643
Epoch 8/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6543 - loss: 0.6414
Epoch 9/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6445 - loss: 0.6528
Epoch 10/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6602 - loss: 0.6373
Epoch 11/150
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6445 -

**Análise das Melhorias:**
A normalização reduz disparidades de escala, a regularização L2 e o dropout combatem overfitting, BatchNormalization estabiliza a distribuição interna das ativações e o EarlyStopping evita treinamento excessivo. A combinação tende a aumentar a curácia média e/ou reduzir o desvio padrão, indicando melhor generalização. Compare estes valores com os anteriores para concluir se houve ganho real.

